farmerbot/internal/types.go

```diff
@@ -0,0 +1,194 @@
+package internal
+
+import (
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	zos "github.com/threefoldtech/zos/client"
+	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+// Config is the inputs for configuration for farmerbot
+type Config struct {
+	FarmID             uint32   `yaml:"farm_id"`
+	IncludedNodes      []uint32 `yaml:"included_nodes"`
+	ExcludedNodes      []uint32 `yaml:"excluded_nodes"`
+	NeverShutDownNodes []uint32 `yaml:"never_shutdown_nodes"`
+	Power              power    `yaml:"power"`
+}
+
+type powerState uint8
+
+const (
+	on = powerState(iota)
+	wakingUP
+	off
+	shuttingDown
+)
+
+// Node represents a node type
+type node struct {
+	substrate.Node
+	resources             consumableResources
+	publicIPsUsed         uint64
+	pools                 []pkg.PoolMetrics
+	gpus                  []zos.GPU
+	hasActiveRentContract bool
+	dedicated             bool
+	neverShutDown         bool
+
+	powerState                powerState
+	timeoutClaimedResources   time.Time
+	lastTimePowerStateChanged time.Time
+	lastTimeAwake             time.Time
+	timesRandomWakeUps        int
+}
+
+// NodeOptions represents the options to find a node
+type NodeOptions struct {
```

Comment:
From the comment I thought it was NodeFilterOption correct? because `NodeExclude` isn't a NodeOption, but a NodeFilteringOption, and probably we should rename to `NodesExcluded`